File: frontend/src/services/auditoria.ts
import api from './api'

export type TipoAccion =
  | 'crear'
  | 'actualizar'
  | 'eliminar'
  | 'login'
  | 'logout'
  | 'ver'
  | 'exportar'
  | 'asignar'
  | 'desasignar'
  | 'cambiar_estado'
  | 'subir_archivo'

export type ModuloAuditado =
  | 'auth'
  | 'usuarios'
  | 'proyectos'
  | 'etapas'
  | 'items_trabajo'
  | 'materiales'
  | 'equipos'
  | 'gastos'
  | 'clientes'
  | 'fotos'
  | 'reportes'
  | 'finanzas'
  | 'alertas'
  | 'jornadas'

export interface Auditoria {
  id: string
  usuario_id?: string
  usuario_nombre?: string
  accion: TipoAccion
  modulo: ModuloAuditado
  descripcion: string
  recurso_id?: string
  recurso_tipo?: string
  datos_adicionales?: Record<string, any>
  ip_address?: string
  user_agent?: string
  created_at: string
}

export interface ResumenActividad {
  periodo: {
    desde: string
    hasta: string
  }
  total_acciones: number
  acciones_por_tipo: Record<string, number>
  acciones_por_modulo: Record<string, number>
  usuarios_mas_activos: Array<{
    usuario_id: string
    usuario_nombre: string
    cantidad: number
  }>
}

export interface FiltrosAuditoria {
  usuario_id?: string
  modulo?: ModuloAuditado
  accion?: TipoAccion
  fecha_desde?: string
  fecha_hasta?: string
  recurso_id?: string
  skip?: number
  limit?: number
}

export const auditoriaService = {
  getAuditorias: async (filtros?: FiltrosAuditoria): Promise<Auditoria[]> => {
    const response = await api.get('/auditoria', { params: filtros })
    return response.data
  },

  getAuditoriaUsuario: async (usuarioId: string, limit: number = 50): Promise<Auditoria[]> => {
    const response = await api.get(`/auditoria/usuario/${usuarioId}`, { params: { limit } })
    return response.data
  },

  getAuditoriaRecurso: async (recursoId: string, limit: number = 50): Promise<Auditoria[]> => {
    const response = await api.get(`/auditoria/recurso/${recursoId}`, { params: { limit } })
    return response.data
  },

  getResumenActividad: async (
    fechaDesde?: string,
    fechaHasta?: string
  ): Promise<ResumenActividad> => {
    const response = await api.get('/auditoria/resumen', {
      params: { fecha_desde: fechaDesde, fecha_hasta: fechaHasta }
    })
    return response.data
  },

  getLoginsUsuario: async (usuarioId: string, limit: number = 20): Promise<Auditoria[]> => {
    const response = await api.get(`/auditoria/logins/${usuarioId}`, { params: { limit } })
    return response.data
  },

  getActividadReciente: async (limit: number = 20): Promise<Auditoria[]> => {
    const response = await api.get('/auditoria/reciente', { params: { limit } })
    return response.data
  },

  getTiposAccion: async (): Promise<{ value: string; label: string }[]> => {
    const response = await api.get('/auditoria/tipos')
    return response.data
  },

  getModulos: async (): Promise<{ value: string; label: string }[]> => {
    const response = await api.get('/auditoria/modulos')
    return response.data
  }
}

// Helpers
export const getAccionColor = (accion: TipoAccion): string => {
  const colors: Record<TipoAccion, string> = {
    crear: 'text-green-600 bg-green-100',
    actualizar: 'text-blue-600 bg-blue-100',
    eliminar: 'text-red-600 bg-red-100',
    login: 'text-purple-600 bg-purple-100',
    logout: 'text-gray-600 bg-gray-100',
    ver: 'text-cyan-600 bg-cyan-100',
    exportar: 'text-orange-600 bg-orange-100',
    asignar: 'text-indigo-600 bg-indigo-100',
    desasignar: 'text-pink-600 bg-pink-100',
    cambiar_estado: 'text-yellow-600 bg-yellow-100',
    subir_archivo: 'text-teal-600 bg-teal-100'
  }
  return colors[accion] || 'text-gray-600 bg-gray-100'
}

export const getAccionLabel = (accion: TipoAccion): string => {
  const labels: Record<TipoAccion, string> = {
    crear: 'Creacion',
    actualizar: 'Actualizacion',
    eliminar: 'Eliminacion',
    login: 'Inicio de Sesion',
    logout: 'Cierre de Sesion',
    ver: 'Visualizacion',
    exportar: 'Exportacion',
    asignar: 'Asignacion',
    desasignar: 'Desasignacion',
    cambiar_estado: 'Cambio de Estado',
    subir_archivo: 'Subida de Archivo'
  }
  return labels[accion] || accion
}

export const getModuloLabel = (modulo: ModuloAuditado): string => {
  const labels: Record<ModuloAuditado, string> = {
    auth: 'Autenticacion',
    usuarios: 'Usuarios',
    proyectos: 'Proyectos',
    etapas: 'Etapas',
    items_trabajo: 'Items de Trabajo',
    materiales: 'Materiales',
    equipos: 'Equipos',
    gastos: 'Gastos',
    clientes: 'Clientes',
    fotos: 'Fotos',
    reportes: 'Reportes',
    finanzas: 'Finanzas',
    alertas: 'Alertas',
    jornadas: 'Jornadas'
  }
  return labels[modulo] || modulo
}

File: backend/app/schemas/auditoria.py
"""Schemas de auditoría."""
from pydantic import BaseModel
from typing import Optional, Dict, Any, List
from uuid import UUID
from datetime import datetime, date
from enum import Enum


class TipoAccion(str, Enum):
    crear = "crear"
    actualizar = "actualizar"
    eliminar = "eliminar"
    login = "login"
    logout = "logout"
    ver = "ver"
    exportar = "exportar"
    asignar = "asignar"
    desasignar = "desasignar"
    cambiar_estado = "cambiar_estado"
    subir_archivo = "subir_archivo"


class ModuloAuditado(str, Enum):
    auth = "auth"
    usuarios = "usuarios"
    proyectos = "proyectos"
    etapas = "etapas"
    items_trabajo = "items_trabajo"
    materiales = "materiales"
    equipos = "equipos"
    gastos = "gastos"
    clientes = "clientes"
    fotos = "fotos"
    reportes = "reportes"
    finanzas = "finanzas"
    alertas = "alertas"
    jornadas = "jornadas"


class AuditoriaBase(BaseModel):
    """Base para auditoría."""
    accion: TipoAccion
    modulo: ModuloAuditado
    descripcion: str
    recurso_id: Optional[UUID] = None
    recurso_tipo: Optional[str] = None
    datos_adicionales: Optional[Dict[str, Any]] = None


class AuditoriaResponse(AuditoriaBase):
    """Respuesta de auditoría."""
    id: UUID
    usuario_id: Optional[UUID] = None
    usuario_nombre: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ResumenActividadResponse(BaseModel):
    """Resumen de actividad del sistema."""
    periodo: Dict[str, str]
    total_acciones: int
    acciones_por_tipo: Dict[str, int]
    acciones_por_modulo: Dict[str, int]
    usuarios_mas_activos: List[Dict[str, Any]]


class FiltrosAuditoria(BaseModel):
    """Filtros para búsqueda de auditorías."""
    usuario_id: Optional[UUID] = None
    modulo: Optional[ModuloAuditado] = None
    accion: Optional[TipoAccion] = None
    fecha_desde: Optional[date] = None
    fecha_hasta: Optional[date] = None
    recurso_id: Optional[UUID] = None
    skip: int = 0
    limit: int = 50
